src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -1489,15 +1527,13 @@ public Object popNotEmpty(DynamicObject array, int n,
                 @Cached("createBinaryProfile()") ConditionProfile minProfile) {
             final int size = strategy.getSize(array);
             final int numPop = minProfile.profile(size < n) ? size : n;
-            final ArrayMirror store = strategy.newMirror(array);
+            final ArrayMirror store = strategy.makeStorageShared(array);
 
             // Extract values in a new array
             final ArrayMirror popped = store.extractRange(size - numPop, size);
 
-            // Null out the popped values from the store
-            final ArrayMirror filler = strategy.newArray(numPop);
-            filler.copyTo(store, 0, size - numPop, numPop);
-            setSize(array, size - numPop);
+            // Remove the end from the original array.
+            setStoreAndSize(array, store.extractRange(0, size - numPop).getArray(), size - numPop);
```

Comment:
We should still null out the removed values here ideally.
What does it gain to use a shared storage for `pop`? Should we unshare the storage instead for `pop`?

Reply:
I've had a think about this. If the storage is already shared then I think it's sensible to simply use `extractRange` to chop the end off, the likely hood is something else can still access the elements, but if the storage is not shared then I agree we should null out the elements.

This is one area where I think having two specializations probably is the right thing.